src/anndata/_io/specs/lazy_methods.py

```diff
@@ -0,0 +1,194 @@
+from __future__ import annotations
+
+from contextlib import contextmanager
+from functools import singledispatch
+from pathlib import Path, PurePosixPath
+from types import MappingProxyType
+from typing import TYPE_CHECKING, overload
+
+import h5py
+import numpy as np
+from scipy import sparse
+
+import anndata as ad
+from anndata.compat import H5Array, H5Group, ZarrArray, ZarrGroup
+
+from .registry import _LAZY_REGISTRY, IOSpec
+
+if TYPE_CHECKING:
+    from collections.abc import Mapping
+    from typing import Any, Literal
+
+    from .registry import Reader
+
+
+@overload
+def make_block_indexer(
+    *,
+    is_csc: Literal[True],
+    stride: int,
+    shape: tuple[int, int],
+    block_id: tuple[int, int],
+) -> tuple[slice, slice]: ...
+@overload
+def make_block_indexer(
+    *,
+    is_csc: Literal[False],
+    stride: int,
+    shape: tuple[int, int],
+    block_id: tuple[int, int],
+) -> tuple[slice]: ...
+
+
+def make_block_indexer(
+    *, is_csc: bool, stride: int, shape: tuple[int, int], block_id: tuple[int, int]
+) -> tuple[slice, slice] | tuple[slice]:
+    index1d = slice(
+        block_id[is_csc] * stride,
+        min((block_id[is_csc] * stride) + stride, shape[is_csc]),
+    )
+    if is_csc:
+        return (slice(None), index1d)
+    return (index1d,)
+
+
+@contextmanager
+def maybe_open_h5(path_or_group: Path | ZarrGroup, elem_name: str):
+    if not isinstance(path_or_group, Path):
+        yield path_or_group
+        return
+    file = h5py.File(path_or_group, "r")
+    try:
+        yield file[elem_name]
+    finally:
+        file.close()
+
+
+_DEFAULT_STRIDE = 1000
```

Comment:
Ok I just spent a long time trying to debug why a zarr store was being overfetched, and it turns out it wasn't.  This number is just crazy high.  So I'm not entirely sure what the best way forward is.  This number is (probably) good for bulk tasks like filtering or PCA but absolutely useless for random access.  For example, on a CSC matrix, you end up fetching 5% of the dataset if there are 20,000 genes.  So we need to be VERY careful here.  Not sure what the answer is, actually.  I don't think exposing `stride` is great because it's specific to csr/csc and also a subset of the functionality of `chunks`, but the flipside is that resolving the size of a sparse matrix requires looking at `attrs`.

Reply:
Exposing a per-matrix (i.e., `layers`) chunking is tough with a monolithic API https://github.com/scverse/anndata/pull/1247 but a monolithic API (i.e., `read_backed`) makes it easier to deal with telling people how to do things, which is currently very difficult.  I'll need to think about that PR.

Reply:
But with that PR, it doesn't really make sense to expose a method called `read_elem_as_xarray` because that's way too specific - only our dataframes work there.  So I'm not sure what the answer is.  Perhaps changing this to `read_lazy`  or something is not a bad idea.

Reply:
I think this is probably a fine value for now. And agree on not exposing `stride`. Also don't know what the answer is.